modules/imap/handler_modules.php

```diff
@@ -550,18 +550,31 @@ public function process() {
                         $this->out('list_filter', $this->request->get['filter']);
                     }
                 }
+                
+                // Handle folder conversion - both IMAP and EWS use hex2bin but for different purposes
+                // IMAP: hex2bin("494e424f58") -> "INBOX" (folder name)
+                // EWS: hex2bin("41414d6b...") -> "AAMkADk5..." (base64 Exchange folder ID)
                 $folder = hex2bin($parts[2]);
+                
                 if (!empty($details)) {
                     if (array_key_exists('folder_label', $this->request->get)) {
                         $folder = $this->request->get['folder_label'];
                         $this->out('folder_label', $folder);
                     }
-                    else {
-                        $folder = hex2bin($parts[2]);
-                    }
+                    
                     $mailbox = Hm_IMAP_List::get_mailbox_without_connection($details);
                     $label = $mailbox->get_folder_name($folder);
-                    $title = array(strtoupper($details['type'] ?? 'IMAP'), $details['name'], $label);
+                    
+                    // For EWS, if we can't get a friendly name, try to use a connected mailbox
+                    if (!$label && isset($details['type']) && $details['type'] === 'ews') {
+                        $connected_mailbox = Hm_IMAP_List::get_connected_mailbox($parts[1], $this->cache);
+                        if ($connected_mailbox && $connected_mailbox->authed()) {
+                            $folder_status = $connected_mailbox->get_folder_status($folder, false);
```

Comment:
This is rather slow to get to the server for each folder label. I think I implemented label caching and returning it immediately when requested in the mailbox class. Also, abstract these types of checks in the mailbox class. We shouldn't pollute handler modules with if statements that depend on server type.